sql: allow REGIONAL BY ROW without experimental flag

otan opened the pull request:
No longer require experimental_enable_implicit_column_partitioning when
creating REGIONAL BY ROW tables.

Resolves https://github.com/cockroachdb/cockroach/issues/59629

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60465)
<!-- Reviewable:end -->

ajstorm reviewed (commented):
Exciting!

Mostly nits below.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60465#-:-MTGMCNUExQjFPtyC810:b-xhp7so)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/alter_index.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60465#-MTGMX74AEGCo3gkPZKD:-MTGMX74AEGCo3gkPZKE:b497ot4) ([raw file](https://github.com/cockroachdb/cockroach/blob/87a82447fea1b853e8c39dbe227b8a5cae992064/pkg/sql/alter_index.go#L89)):*
> ```Go
> 				t.PartitionBy,
> 				nil, /* allowedNewColumnNames */
> 				params.p.EvalContext().SessionData.ImplicitColumnPartitioningEnabled ||
> ```

This might be easier to read if we pulled it out into a separate variable above:

```allowImplicitPartitioning :=  params.p.EvalContext().SessionData.ImplicitColumnPartitioningEnabled || n.tableDesc.IsLocalityRegionalByRow()```

We could then add a comment describing why we allow implicit partitioning in these cases.
___
*[pkg/sql/alter_primary_key.go, line 314 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60465#-MTGNPsR16BMy27CTpkQ:-MTGNPsR16BMy27CTpkR:b-filbxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/87a82447fea1b853e8c39dbe227b8a5cae992064/pkg/sql/alter_primary_key.go#L314)):*
> ```Go
> 			partitionAllBy,
> 			allowedNewColumnNames,
> 			// PARTITION ALL BY is only enabled if the experimental flag was previously set.
> ```

I find this comment (and the one below) confusing.  Seems like we're unconditionally setting it to true here.  What am I missing?
___
*[pkg/sql/create_table.go, line 1621 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60465#-MTGNInK67z5FfCLht6q:-MTGNInK67z5FfCLht6r:b9u6v0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/87a82447fea1b853e8c39dbe227b8a5cae992064/pkg/sql/create_table.go#L1621)):*
> ```Go
> 	}
> 
> 	allowImplicitPartitioning := (sessionData != nil && sessionData.ImplicitColumnPartitioningEnabled) ||
> ```

Ahh, you did it down here.  Can we do this everywhere?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60465)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm)

___
*[pkg/sql/alter_primary_key.go, line 314 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60465#-MTGNPsR16BMy27CTpkQ:-MTI0pthCF2lm99bzhDs:b-wf6880) ([raw file](https://github.com/cockroachdb/cockroach/blob/87a82447fea1b853e8c39dbe227b8a5cae992064/pkg/sql/alter_primary_key.go#L314)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

I find this comment (and the one below) confusing.  Seems like we're unconditionally setting it to true here.  What am I missing?
</blockquote></details>

I've refined it a little more.
Basically if a session had the experimental flag to true, but we do alter primary key later, we expect it to "carry over".


<!-- Sent from Reviewable.io -->

otan committed:
sql: allow REGIONAL BY ROW without experimental flag

No longer require experimental_enable_implicit_column_partitioning when
creating REGIONAL BY ROW tables.

Release note: None

ajstorm reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60465#-:-MTLg3T682eiBYKTEyVu:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/alter_primary_key.go, line 314 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60465#-MTGNPsR16BMy27CTpkQ:-MTLfzF4CCnekEDcdZxA:bt4d9ep) ([raw file](https://github.com/cockroachdb/cockroach/blob/87a82447fea1b853e8c39dbe227b8a5cae992064/pkg/sql/alter_primary_key.go#L314)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I've refined it a little more.
Basically if a session had the experimental flag to true, but we do alter primary key later, we expect it to "carry over".
</blockquote></details>

Looks good to me.  Much clearer.  Thx!


<!-- Sent from Reviewable.io -->

otan commented:
thanks

bors r=ajstorm
